### PR TITLE
Bytecode generation

### PR DESCRIPTION
**Pull request by isomethane**

### CODE REVIEW

File: reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/Util.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+}
+
+fun InstructionAdapter.checkCast(castType: Type, checkForNull: Boolean) {
+    if (checkForNull) {
+        val nonNullLabel = Label()
+
+        dup()
+        ifnonnull(nonNullLabel)
+        anew(Type.getObjectType("kotlin/TypeCastException"))
```
Comment (nbirillo) [inline reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/Util.kt:28]: You use different consts like `kotlin/TypeCastException` for paths. I think we can create storage for them or something like this. I mean, if you use string constant you can make a mistake. But if we use variables, we are sure the same variable has the same value in the different places

File: reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/ReflektGeneratorExtension.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        get() = when (terminalFunctionName) {
+            ReflektTerminalFunctionName.TO_LIST -> InstructionAdapter::invokeListOf
+            ReflektTerminalFunctionName.TO_SET -> InstructionAdapter::invokeSetOf
+        }
+
+    fun getUses(uses: ReflektUses, invokeArguments: SubTypesToAnnotations, typeMapper: KotlinTypeMapper): List<Type> =
+        when (val type = name) {
+            ReflektName.OBJECTS, ReflektName.CLASSES -> {
+                val items = if (type == ReflektName.OBJECTS) uses.objects else uses.classes
+                items.getValue(invokeArguments).map {
```
Comment (nbirillo) [inline reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/ReflektGeneratorExtension.kt:64]: getValue throws errors if the key does not exist. Is it ok? Let's use 
```
items[invokeArguments]?.map {
                    ...
                } ?: emptyList()
```
or
```
items[invokeArguments]?.map {
                    ...
                } ?: error("Some error message")
```

File: reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/ReflektGeneratorExtension.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
+package io.reflekt.plugin.generation.bytecode
```
Comment (nbirillo) [inline reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/ReflektGeneratorExtension.kt:1]: I think we should add some comments to this file, especially in `ReflektInvokeNames` class.

File: reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/ReflektGeneratorExtension.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import org.jetbrains.kotlin.codegen.asmType
+import org.jetbrains.kotlin.codegen.extensions.ExpressionCodegenExtension
+import org.jetbrains.kotlin.codegen.state.KotlinTypeMapper
+import org.jetbrains.kotlin.psi.synthetics.findClassDescriptor
+import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.jetbrains.org.objectweb.asm.Type
+import org.jetbrains.org.objectweb.asm.commons.InstructionAdapter
+
+class ReflektGeneratorExtension(private val messageCollector: MessageCollector? = null) : ExpressionCodegenExtension {
+    override fun applyFunction(receiver: StackValue, resolvedCall: ResolvedCall<*>, c: ExpressionCodegenExtension.Context): StackValue? {
```
Comment (nbirillo) [inline reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/ReflektGeneratorExtension.kt:23]: Please, use `messageCollector` to log some intermediate information to make debugging easier, for example before `return null` and etc

**Review by nbirillo — CHANGES_REQUESTED** [review]
Great!

**Review by nbirillo — COMMENTED** [review]
Thanks a lot, it looks great! I have one question: can we add several tests for the bytecode generation to be sure that the generated code is correct? We can do it in another PR of course. What do you think?

**Comment by isomethane** [comment]
Yes, I will add tests.

**Review by nbirillo — APPROVED** [review]
Cool! Let's add tests in the next PR

**Pull request merged**